proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,671 @@
+``($)`` will be:
+
+::
+
+  ($) :: (a:p -> b) ⊸ a:p -> b
+
+The precise content of the library is out of scope of this proposal,
+but it will also contain convenient types to work with linear types
+such as:
+
+::
+
+   data Unrestricted a where
+     Unrestricted :: a -> Unrestricted a
+
+Formalism
+~~~~~~~~~
+
+This section describes the changes required in Core.
+
+For ease of type-checking, we add another multiplicity: ``0``
+representing definite absence of consumption of an argument (``0``
+has also been used by `Conor McBride
+<https://link.springer.com/chapter/10.1007/978-3-319-30936-1_12>`_ to
+handle dependent types, which may matter for Dependent Haskell).
+
+The only requirement on multiplicities is that they form a
+sup-semi-lattice-ordered semi-ring. That is: there is a sum and a
+product with the usual distributivity laws, a (computable) order
+compatible with the sum and product, such that each pair of
+multiplicities has a (computable) join. Even if there is only three
+multiplicities in this proposal, the proposal is structured to allow
+future extensions.
+
+Variables are added and multiplied symbolically. Therefore
+multiplicity expressions are multi-variate polynomials in the
+multiplicity semi-ring.
+
+In Core, every variable is labelled, with its multiplicity (just like
+it is with its type prior to this proposal). This multiplicity is used
+to infer the multiplicity in the type of functions.
+
+In order to cope with the fact that
+
+::
+
+  fst :: (a, b) -> a
+  fst (a, _) = a
+
+is well-typed but
+
+::
+
+  fst :: (a, b) ->. a
+  fst (a, _) = a
+
+isn't, ``case`` expressions are annotated with a multiplicity as
+well. This multiplicity scales the multiplicity of constructors'
+fields. The latter example is elaborated into a ``case_1`` so the
+``_`` pattern is linear, which is prohibited, in the former we have a
+``case_ω`` so the multiplicity of both fields are scaled by ``ω`` (in
+particular ``_`` is an unrestricted pattern) and the expression
+typechecks.
+
+This has one consequences: ``case_0`` cannot be allowed, as it would
+break the definition of ``0`` (it would force something which is, by
+definition, definitely not consumed, for instance, it would allow
+computing the length of a list with multiplicity ``0``). But we do
+want to accept ``case_p`` when ``p`` is a variable. Therefore me must
+take the convention that variables never stand for the ``0``
+multiplicity, and in particular that ``0`` is not a valid argument
+for a multiplicity application.
+
+Remark: ``let`` binders are decorated like ``case``, with the
+restriction that recursive ``let`` binders are necessarily decorated
+with ``ω``.
+
+Effect and Interactions
+-----------------------
+
+A staple of this proposal is that it does not modify Haskell for those
+who don't want to use it, or don't know of linear types. Even if an
+API exports linear types, they are easy to ignore: just imagine that
+the arrows are regular arrows, it will work as expected.
```

Comment:
This is true only if the exported types are multiplicity polymorphic. `map :: (a ->. b) -> [a] ->. [b]` would certainly reject `map f` if `f :: a -> b`.

Reply:
It would. Neither of these types is more general than the other.

Reply:
In that case, linear types are not as easy to ignore as the text in the proposal says.